### ISSUE TITLE
Spock breaks try-catch logic when testing Java code

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sampig/spock#166]
```
Spock 0.5-groovy-1.7, Java 1.6.

I am using a Spock specification to unit test a Java class. The class under 
test wraps IOExceptions up in a custom Exception, which extends IOException.

For example, a method in the class under test looks something like:

public Response get(URI uri) {
    try {
        Response response = client.get(uri)
    } catch (IOException e) {
        throw new CustomException(e)
    }
    /* Do something with the response... */
    return response;
}

I am testing the exception handling using a feature method similar to this:

def "IOException is wrapped in CustomException" () {
    setup:
        def client = Mock(Client)
    when:
        classUnderTest.get(uri)
    then:
        1 * client.get(_ as URI) >> { throw new IOException() }
        CustomException e = thrown()
        e.cause == IOException
}

However, this feature method fails, saying that an IOException was thrown 
rather than a CustomException.  If I add the following catch block to the 
try-catch in the class under test:

catch (Throwable t) {
    System.out.println(t.getClass().getCanonicalName());
    throw new RuntimeException(t);
}

then I find that this catch block is executed, and the throwable is of type 
org.codehaus.groovy.runtime.InvokerInvocationException, which explains why the 
IOException block was not executed.

Is there any way to make Spock work when testing exception handling in Java 
code?

```

Original issue reported on code.google.com by `Ewan.Dawson` on 10 Jan 2011 at 11:35

**Comment by GoogleCodeExporter** [comment]
```
I'm not aware of any problems with testing exception handling, and I cannot 
reproduce your problem. Which Groovy version are you using? Can you provide a 
reproducible example? By the way, "e.cause == IOException" should be "e.cause 
instanceof IOException".
```

Original comment by `pnied...@gmail.com` on 11 Jan 2011 at 3:51

**Comment by GoogleCodeExporter** [comment]
```
Attached is example which reproduces this problem. I am using Groovy 1.7.5.

Cheers,
Ewan
```

Original comment by `Ewan.Dawson` on 11 Jan 2011 at 9:17

Attachments:
- [spock-exception-test.zip](https://storage.googleapis.com/google-code-attachments/spock/issue-166/comment-2/spock-exception-test.zip)

**Comment by GoogleCodeExporter** [comment]
```
Thanks. I can reproduce the problem now. It seems to be somehow related to the 
usage of a mock object, but I have to investigate further.
```

Original comment by `pnied...@gmail.com` on 12 Jan 2011 at 10:01
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
This problem is caused by http://jira.codehaus.org/browse/GROOVY-4662. I've 
implemented a workaround that fixes the problem.
```

Original comment by `pnied...@gmail.com` on 15 Feb 2011 at 2:41
- Changed state: **Fixed**
- Added labels: **Milestone-0.6**

**Comment by GoogleCodeExporter** [comment]
```
Hey Guys, I've run into this very same issue (Spock 0.5, Groovy 1.8).  If there 
is a workaround as alluded to above, could you share it?  Thanks
```

Original comment by `prysta...@gmail.com` on 5 Mar 2012 at 2:09

**Comment by GoogleCodeExporter** [comment]
```
Already fixed in 0.6.
```

Original comment by `pe...@gradle.biz` on 5 Mar 2012 at 2:11
